src/amici.cpp

```diff
@@ -1383,6 +1384,375 @@ int getDiagnosisB(int it, void *ami_mem, UserData *udata, ReturnData *rdata, Tem
+/* ------------------------------------------------------------------------------------- */
+
+
+int applyNewtonsMethod(void *ami_mem, UserData *udata, ReturnData *rdata, TempData *tdata, int status, int ntry) {
+    /**
+     * applyNewtonsMethod applies Newtons method to the current state x to find the steady state
+     *
+     * @param[in] ami_mem pointer to the solver memory block @type *void
+     * @param[in] udata pointer to the user data struct @type UserData
+     * @param[in] tdata pointer to the temporary data struct @type TempData
+     * @param[out] tdata pointer to the temporary data struct @type TempData
+     * @param[out] rdata pointer to the return data struct @type ReturnData
+     * @return void
+     */
+
+    int ix;
+    int inewt = 0;
+    int stat1 = -1;
+    
+    double res;
+    double rel_res;
+    double res_old;
+    double gamma = 1.0;
+    realtype *x_tmp;
+    
+    N_Vector delta = N_VNew_Serial(udata->nx);
+    N_Vector rel_x = N_VNew_Serial(udata->nx);
+    N_Vector tmp_x = N_VNew_Serial(udata->nx);
+    
+    N_VConst(0.0, delta);
+    
+    // Check, how fxdot is used exactly within AMICI...
+    fxdot(tdata->t, tdata->x, tdata->dx, tdata->xdot, udata);
+    res = sqrt(N_VDotProd(tdata->xdot,tdata->xdot));
+    
+    // Check for relative error, but make sure not to divide by 0!
+    // Ensure positivity of the state
+    N_VScale(1.0, tdata->x, tmp_x);
+    N_VAbs(tmp_x, tmp_x);
+    x_tmp = N_VGetArrayPointer(tmp_x);
+    for (ix=0; ix<udata->nx; ix++) {
+        if (x_tmp[ix] < udata->atol) {
+            x_tmp[ix] = udata->atol;
+        }
+    }
+    N_VDiv(tdata->xdot, tmp_x, rel_x);
+    rel_res = sqrt(N_VDotProd(rel_x, rel_x));
+    if (res < udata->atol || rel_res < udata->rtol) {
+        // Clean up worksapce
+        N_VDestroy_Serial(delta);
+        N_VDestroy_Serial(rel_x);
+        N_VDestroy_Serial(tmp_x);
+        
+        // Return Success;
+        return (0);
+    }
+    
+    // compute the inital search direction
+    status = getNewtonStep(udata, rdata, tdata, ami_mem, ntry, inewt, delta);
+    
+    // Copy the current state to the old one, make up a new vector for Jdiag
+    N_VScale(1.0, tdata->x, tdata->x_old);
+    N_VScale(1.0, tdata->xdot, tdata->xdot_old);
+    
+    // Newton iterations
+    for(inewt=0; inewt<udata->newton_maxsteps; inewt++) {
+        
+        // Try a full, undamped Newton step
+        N_VLinearSum(1.0, tdata->x_old, gamma, delta, tdata->x);
+        
+        // Ensure positivity of the state
+        x_tmp = N_VGetArrayPointer(tdata->x);
+        for (ix=0; ix<udata->nx; ix++) {
+            if (x_tmp[ix] < 0.0) {
+                x_tmp[ix] = 0.0;
+            }
+        }
+        
+        // Compute new xdot
+        fxdot(tdata->t, tdata->x, tdata->dx, tdata->xdot, udata);
+        
+        // Compute residuals
+        res = N_VDotProd(tdata->xdot, tdata->xdot);
+        res_old = N_VDotProd(tdata->xdot_old, tdata->xdot_old);
```

Comment:
Does `res_old` need to be recomputed? Can't we use the one from the previous iteration?